### PR TITLE
ML - Verifying ODBC Driver

### LINKED ISSUE #10694: ML - refactor and add test for Python and R package management provider
- Azure Data Studio Version:

the code for package management provider is not readable and can cause bug by changing the code easily. Need to refactor and add more tests

### PR DESCRIPTION
**Pull request by llali**
Verifying ODBC driver 17 before package management. Showing an error message to user with a "learn more" link so user would get more info about how to install ODBC.

also fixes #10694

### CODE REVIEW

File: extensions/machine-learning/src/common/utils.ts [hunk 1]
```diff
@@ -283,3 +283,34 @@ export function getPythonExeName(): string {
 export function getUserHome(): string | undefined {
 	return process.env.HOME || process.env.USERPROFILE;
 }
+
+export function getKeuValueString(key: string, value: string, separator: string = '='): string {
```
Comment (Benjin) [inline extensions/machine-learning/src/common/utils.ts:287]: typo: keyValue

File: extensions/machine-learning/src/common/utils.ts [hunk 2]
```diff
@@ -283,3 +283,34 @@ export function getPythonExeName(): string {
...
+	}
+	let index = connection.serverName.indexOf(',');
+	if (index > 0) {
+		return connection.serverName.substring(index + 1);
+	} else {
+		return '1433';
+	}
+}
+
+export function getServeName(connection: azdata.connection.ConnectionProfile): string {
```
Comment (Benjin) [inline extensions/machine-learning/src/common/utils.ts:303]: typo: serve_r_

File: extensions/machine-learning/src/common/utils.ts [hunk 3]
```diff
@@ -283,3 +283,34 @@ export function getPythonExeName(): string {
...
+export function getServerPort(connection: azdata.connection.ConnectionProfile): string {
+	if (!connection) {
+		return '';
+	}
+	let index = connection.serverName.indexOf(',');
+	if (index > 0) {
+		return connection.serverName.substring(index + 1);
+	} else {
+		return '1433';
+	}
```
[the remark this answers is not in the document]
Comment (llali) [reply extensions/machine-learning/src/common/utils.ts:300]: I don't think the way we set port in server name, localhost,1433 is what Uri can parse. am I wrong?


---
In reply to: [463757200](https://github.com/microsoft/azuredatastudio/pull/11587#discussion_r463757200) [](ancestors = 463757200)

File: extensions/machine-learning/src/common/constants.ts [hunk 4]
```diff
@@ -48,6 +49,7 @@ export const adsPythonBundleVersion = '0.0.1';
 export const msgYes = localize('msgYes', "Yes");
 export const msgNo = localize('msgNo', "No");
 export const managePackageCommandError = localize('mls.managePackages.error', "Package management is not supported for the server. Make sure you have Python or R installed.");
+export const verifyOdbcDriverError = localize('mls.verifyOdbcDriverError.error', "'{0}' is required  for package management. Please make sure it is installed and setup correctly.", supportedODBCDriver);
```
Comment (udeeshagautam) [inline extensions/machine-learning/src/common/constants.ts:52]: "on server" right - lets have that clear in string?

**Comment by udeeshagautam** [comment]
looks good to me. @alanrenmsft tagging you also <same thing as other PR>

File: extensions/machine-learning/src/common/constants.ts (hunk 4 above)
Comment (Benjin) [inline extensions/machine-learning/src/common/constants.ts:52]: Nit: "set up" is two words when used as a verb

File: extensions/machine-learning/src/common/utils.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Benjin) [reply extensions/machine-learning/src/common/utils.ts:300]: oh yeah, not with a comma, I think.

File: extensions/machine-learning/src/common/constants.ts (hunk 4 above)
Comment (llali) [reply extensions/machine-learning/src/common/constants.ts:52]: no locally


---
In reply to: [468935412](https://github.com/microsoft/azuredatastudio/pull/11587#discussion_r468935412) [](ancestors = 468935412)

**Pull request merged**